Author: Apollon Oikonomopoulos <apoikos@debian.org>
Date: Tue, 19 Sep 2017 14:56:25 +0300
Bug-Debian: https://bugs.debian.org/866752
Description: Handle requests to enable/disable deprecated SSL versions

SSL_TXT_SSLV2 is gone from OpenSSL, and most likely SSL_TXT_SSLV3 will
also be removed one day. Handle these removals in a graceful manner:

 - Do not complain about unknown protocols
 - Warn at config parse time about their use
---
 src/config/config-parser.c                     | 25 +++++++++++++++++++++++++
 src/lib-ssl-iostream/iostream-openssl-common.c |  8 ++++++++
 2 files changed, 33 insertions(+)

--- a/src/config/config-parser.c
+++ b/src/config/config-parser.c
@@ -23,6 +23,9 @@
 #ifdef HAVE_GLOB_H
 #  include <glob.h>
 #endif
+#ifdef HAVE_OPENSSL
+#  include <openssl/ssl.h>
+#endif
 
 #ifndef GLOB_BRACE
 #  define GLOB_BRACE 0
@@ -454,6 +457,29 @@
 			ssl_warned = TRUE;
 		}
 
+#if defined(HAVE_OPENSSL) && (!defined(SSL_TXT_SSLV2) || !defined(SSL_TXT_SSLV3))
+		{
+			const char *ssl_protocols;
+			const char *const *tmp;
+			ssl_protocols = get_str_setting(parsers[i], "ssl_protocols", "");
+			tmp = t_strsplit_spaces(ssl_protocols, ", ");
+			for (; *tmp != NULL; tmp++) {
+				const char *name = *tmp;
+				if (*name == '!')
+					name++;
+#  ifndef SSL_TXT_SSLV2
+				if (strcasecmp(name, "SSLv2") == 0)
+					i_warning("SSLv2 not supported by OpenSSL. "
+						  "Please consider removing it from ssl_protocols.");
+#  endif
+#  ifndef SSL_TXT_SSLV3
+				if (strcasecmp(name, "SSLv3") == 0)
+					i_warning("SSLv3 not supported by OpenSSL. "
+						  "Please consider removing it from ssl_protocols.");
+#  endif
+			}
+		}
+#endif
 		ret = config_filter_parser_check(ctx, tmp_parsers, error_r);
 		config_filter_parsers_free(tmp_parsers);
 		p_clear(tmp_pool);
--- a/src/lib-ssl-iostream/iostream-openssl-common.c
+++ b/src/lib-ssl-iostream/iostream-openssl-common.c
@@ -38,11 +38,19 @@
 		if (strcasecmp(name, SSL_TXT_SSLV2) == 0)
 			proto = DOVECOT_SSL_PROTO_SSLv2;
 		else
+#else
+		if (strcasecmp(name, "SSLv2") == 0)
+			;
+		else
 #endif
 #ifdef SSL_TXT_SSLV3
 		if (strcasecmp(name, SSL_TXT_SSLV3) == 0)
 			proto = DOVECOT_SSL_PROTO_SSLv3;
 		else
+#else
+		if (strcasecmp(name, "SSLv3") == 0)
+			;
+		else
 #endif
 		if (strcasecmp(name, SSL_TXT_TLSV1) == 0)
 			proto = DOVECOT_SSL_PROTO_TLSv1;
